### PR TITLE
Deep spin new

### PR DESCRIPTION
**Pull request by hztttt**

### CODE REVIEW

File: deepmd/descriptor/se_a.py [hunk 1]
```diff
@@ -275,6 +289,30 @@ def get_ntypes(self) -> int:
...
+        """
+        Returns the list of magnitude of atomic spin for each atom type
+        """
+        return self.spin_norm
+
+    def get_virtual_len(self) -> List[float]:
+        """
+        Returns the list of distance between real atom and virtual atom for each atom type
+        """
+        return self.virtual_len
```
Comment (njzjz) [inline deepmd/descriptor/se_a.py:314]: Instead of creating the methods here, I suggest creating a new class `Spin`.

```py
class Spin:
    def __init__(self, ...):
        ...

    def get_ntypes_spin(self) -> int:
        ...
```

Then pass a class to the descriptor:
```py
spin = Spin(**spin_param)
descrpt_param['spin'] = spin
```

you can call
```py
self.spin.get_ntypes_spin()
```

Then we don't need to repeat it for every descriptor, and we are much easier to implement it for other descriptors.

File: deepmd/infer/deep_pot.py [hunk 2]
```diff
@@ -190,15 +191,33 @@ def __init__(
...
+            self.rcut,
+            self.dfparam,
+            self.daparam,
+            self.tmap,
+        ] = run_sess(
             self.sess,
-            [self.t_ntypes, self.t_rcut, self.t_dfparam, self.t_daparam, self.t_tmap],
+            [
+                self.t_ntypes,
+                self.t_ntypes_spin,
```
Comment (njzjz) [inline deepmd/infer/deep_pot.py:205]: This will break other graphs which do not have this tensor.

File: deepmd/loss/ener.py [hunk 3]
```diff
@@ -252,6 +252,340 @@ def eval(self, sess, feed_dict, natoms):
         return results
 
 
+class EnerSpinLoss(Loss):
```
Comment (njzjz) [inline deepmd/loss/ener.py:255]: It looks that this class just add `use_spin` compared to `EnerLoss`. Why not directly add this option to `EnerLoss`?

File: examples/NiO/data/data_1/type.raw [hunk 4]
```diff
@@ -0,0 +1 @@
+0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
```
Comment (njzjz) [inline examples/NiO/data/data_1/type.raw:1]: The number of data files is too much. As it is a simple example, one set is enough.

File: examples/NiO/lmp/0.dump [hunk 5]
```diff
@@ -0,0 +1,13 @@
+ITEM: TIMESTEP
+0
+ITEM: NUMBER OF ATOMS
+4
+ITEM: BOX BOUNDS xy xz yz pp pp pp
+0.0000000000000000e+00 1.3657940670000000e+01 4.2674213200000004e+00
+0.0000000000000000e+00 4.1221753799999998e+00 4.2674488400000001e+00
+0.0000000000000000e+00 2.5250889600000002e+00 1.2877270000000001e+00
+ITEM: ATOMS id x y z c_spin[1] c_spin[2] c_spin[3] fx fy fz c_spin[5] c_spin[6] c_spin[7]
```
Comment (njzjz) [inline examples/NiO/lmp/0.dump:9]: log files should be removed.

File: examples/NiO/lmp/lmp.sh [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+#SBATCH -J srs-lmp
+#SBATCH --gpus 1
+
+# export PATH=/data/run01/scv6266/soft/lammps-DP/src:$PATH
+export PATH=/data/run01/scv6266/soft/lammps_23Jun2022/src:$PATH
```
Comment (njzjz) [inline examples/NiO/lmp/lmp.sh:6]: This specific environment configuration is not required for others.

File: examples/NiO/se_e2_a/input_v2_compat.json [hunk 7]
```diff
@@ -0,0 +1,93 @@
+{
```
Comment (njzjz) [inline examples/NiO/se_e2_a/input_v2_compat.json:1]: This temporary file should be removed.

File: examples/NiO/se_e2_a/checkpoint [hunk 8]
```diff
@@ -0,0 +1,6 @@
+model_checkpoint_path: "model.ckpt-1000000"
```
Comment (njzjz) [inline examples/NiO/se_e2_a/checkpoint:1]: This temporary file should be removed.

File: examples/NiO/se_e2_a/input.json [hunk 9]
```diff
@@ -0,0 +1,98 @@
+{
```
Comment (njzjz) [inline examples/NiO/se_e2_a/input.json:1]: I suggest renaming the directory name from `NiO` to `spin` as it is more meaningful to others.

File: source/api_cc/src/DeepPot.cc [hunk 10]
```diff
@@ -293,6 +294,7 @@ void DeepPot::init(const std::string& model,
   }
   cell_size = rcut;
   ntypes = get_scalar<int>("descrpt_attr/ntypes");
+  ntypes_spin = get_scalar<int>("descrpt_attr/ntypes_spin");
```
Comment (njzjz) [inline source/api_cc/src/DeepPot.cc:297]: This will break other models which do not have this tensor. You may need to use `try...catch`.

File: source/lmp/pair_deepmd.cpp [hunk 11]
```diff
@@ -902,6 +1041,8 @@ static bool is_key(const string &input) {
   keys.push_back("atomic");
   keys.push_back("relative");
   keys.push_back("relative_v");
+  keys.push_back("virtual_len");
+  keys.push_back("spin_norm");
```
Comment (njzjz) [inline source/lmp/pair_deepmd.cpp:1045]: Please add a test to `source/lmp/tests` for new keys.

File: deepmd/infer/deep_pot.py [hunk 12]
```diff
@@ -335,16 +354,17 @@ def eval(
...
-            if atomic:
-                raise RuntimeError("modifier does not support atomic modification")
-            me, mf, mv = self.dm.eval(coords, cells, atom_types)
-            output = list(output)  # tuple to list
-            e, f, v = output[:3]
-            output[0] += me.reshape(e.shape)
-            output[1] += mf.reshape(f.shape)
-            output[2] += mv.reshape(v.shape)
-            output = tuple(output)
+        if self.ntypes_spin == 0:
```
Comment (wanghan-iapcm) [inline deepmd/infer/deep_pot.py:357]: Why do you need the condition here ?  It exclude the case: spin is used with modifier.

File: deepmd/infer/deep_pot.py [hunk 13]
```diff
@@ -498,7 +519,10 @@ def _eval_inner(
             av = np.reshape(av, [nframes, natoms, 9])
             return energy, force, virial, ae, av
         else:
-            return energy, force, virial
+            if self.ntypes_spin == 0:
+                return energy, force, virial
+            else:
+                return energy, force, virial, natoms_vec
```
Comment (wanghan-iapcm) [inline deepmd/infer/deep_pot.py:525]: This is a breaking change of the `DeepPot.eval`. I would not recommend doing so.

Please add a new method to get the `natoms_vec`.

File: deepmd/model/ener.py [hunk 14]
```diff
@@ -268,6 +281,114 @@ def build(
             )
             force = force + sw_force + tab_force
 
+        force = tf.reshape(force, [-1, 3 * natoms[1]])
+        if self.spin is not None:
```
Comment (wanghan-iapcm) [inline deepmd/model/ener.py:285]: Please do not write everything in the body of `build`. Pack all the code in the condition to a new method, and provide a clear interface letting people know what does the method do.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]
- please remove from the examples:
  - all .sh files. If the command is not standard, write the instruction in the manual. 
  - all checkpoint files
  - all model binary files.
  - other files that are not needed to reproduce the training, like lcurve.out, input_v2_compat.json, train.sh, out.json...
  - lammps dumped trajectories.
  - other lammps files that are not needed to reproduce the MD.

File: deepmd/descriptor/se_a.py (hunk 1 above)
Comment (njzjz) [reply deepmd/descriptor/se_a.py:314]: I see `spin` is in the `model` section. I am not sure if it is for descriptors only or like the type embedding.

Comment (hztttt) [reply deepmd/descriptor/se_a.py]: `spin` is used in both `DescrptSeASpin` and `model`. So, We have created class `spin` following @njzjz  suggestions.

File: deepmd/infer/deep_pot.py (hunk 2 above)
Comment (hztttt) [reply deepmd/infer/deep_pot.py]: ok, we have resolved this problem.

File: deepmd/loss/ener.py (hunk 3 above)
Comment (hztttt) [reply deepmd/loss/ener.py:271]: Force is divided into atomic force and magnetic force, and their losses are built individually. So, we need `EnerSpinLoss`.

File: examples/NiO/lmp/0.dump (hunk 5 above)
Comment (hztttt) [reply examples/NiO/lmp/0.dump]: OK.

File: examples/NiO/se_e2_a/input_v2_compat.json (hunk 7 above)
Comment (hztttt) [reply examples/NiO/se_e2_a/input_v2_compat.json]: OK.

File: examples/NiO/se_e2_a/checkpoint (hunk 8 above)
Comment (hztttt) [reply examples/NiO/se_e2_a/checkpoint]: OK.

File: source/lmp/pair_deepmd.cpp (hunk 11 above)
Comment (hztttt) [reply source/lmp/pair_deepmd.cpp:1044]: We have changed source code of LAMMPS. So, We can't provide a unit test directly.

File: source/api_cc/src/DeepPot.cc [hunk 15]
```diff
@@ -56,6 +56,7 @@ static void run_model(
 
   dener = oe(0);
   std::vector<VALUETYPE> dforce(3 * nall);
+  dforce_.resize(nall * 3);
```
[the remark this answers is not in the document]
Comment (hztttt) [reply source/api_cc/src/DeepPot.cc]: We have removed this resize here.

File: deepmd/fit/ener.py [hunk 16]
```diff
@@ -164,6 +167,10 @@ def __init__(
         self.rcond = rcond
         self.seed = seed
         self.uniform_seed = uniform_seed
+        self.use_spin = use_spin
+        self.spin_norm = spin_norm
+        self.virtual_len = virtual_len
+        self.ntypes_spin = descrpt.get_ntypes_spin()
```
[the remark this answers is not in the document]
Comment (hztttt) [reply deepmd/fit/ener.py]: We have handled this error.

File: deepmd/infer/deep_pot.py (hunk 12 above)
Comment (hztttt) [reply deepmd/infer/deep_pot.py]: We didn't use spin with modifier before. But now, we have included this case. But, it needs more tests.

File: deepmd/infer/deep_pot.py (hunk 13 above)
Comment (hztttt) [reply deepmd/infer/deep_pot.py]: We solved this problem.

File: deepmd/fit/ener.py [hunk 17]
```diff
@@ -568,6 +597,8 @@ def build(
             start_index = 0
             outs_list = []
             for type_i in range(self.ntypes):
+                if type_i >= ntypes_atom:
+                    break
```
Comment (njzjz) [inline deepmd/fit/ener.py:601]: It looks like you can directly write `range(ntypes_atom)`.

File: deepmd/fit/ener.py [hunk 18]
```diff
@@ -645,7 +676,7 @@ def build(
         self.atom_ener_before = outs * atype_filter
         self.add_type = tf.reshape(
             tf.nn.embedding_lookup(self.t_bias_atom_e, self.atype_nloc),
-            [tf.shape(inputs)[0], natoms[0]],
+            [tf.shape(inputs)[0], tf.reduce_sum(natoms[2 : 2 + ntypes_atom])],
```
Comment (njzjz) [inline deepmd/fit/ener.py:679]: `natoms[0]` also appears in line 687 and line 691 after this line.

It seems to me that one can directly pass `tf.reduce_sum(natoms[2 : 2 + ntypes_atom])` to `natoms[0]` when preparing the feed dictionary. Is there any problem?

File: examples/spin/lmp/in.force [hunk 19]
```diff
@@ -0,0 +1,47 @@
+#!/bin/sh
```
Comment (njzjz) [inline examples/spin/lmp/in.force:1]: Not clear why this file can be executed by `sh`.

File: source/lmp/pair_deepmd.cpp [hunk 20]
```diff
@@ -1049,8 +1165,21 @@ void PairDeepMD::settings(int narg, char **arg) {
       eps_v = strtof(arg[iarg + 1], NULL);
 #endif
       iarg += 2;
+    } else if (string(arg[iarg]) == string("virtual_len")) {
```
Comment (njzjz) [inline source/lmp/pair_deepmd.cpp:1168]: Please add documentation for new paramters

File: deepmd/utils/spin.py [hunk 21]
```diff
@@ -0,0 +1,43 @@
+from typing import (
+    List,
+)
+
+
+class Spin:
+    """Parameters
```
Comment (njzjz) [inline deepmd/utils/spin.py:7]: Please add a brief for this class on the first line.

**Review by njzjz — CHANGES_REQUESTED** [review]
The current code looks better. I haven't had time to review codes carefully lately, but I commented on some obvious issues.

File: deepmd/fit/ener.py (hunk 18 above)
Comment (hztttt) [reply deepmd/fit/ener.py:679]: For the case with spin, `tf.reduce_sum(natoms[2 : 2 + ntypes_atom])` only counts real atoms; For the case without spin, `tf.reduce_sum(natoms[2 : 2 + ntypes_atom])` is equal to `natoms[0]`.

File: source/lmp/pair_deepmd.cpp (hunk 20 above)
Comment (hztttt) [reply source/lmp/pair_deepmd.cpp:1168]: We don't know where the documentation should be put. For example, we didn't find the documentation for `relative`. Please give us an example.

Comment (njzjz) [reply source/lmp/pair_deepmd.cpp:1168]: https://github.com/deepmodeling/deepmd-kit/blob/7020b345d3973b6de73431938315387aa2cedde7/doc/third-party/lammps-command.md?plain=1#L72

Comment (hztttt) [reply source/lmp/pair_deepmd.cpp:1168]: OK. We have added documentations for spin parameters.

File: deepmd/loss/ener.py [hunk 22]
```diff
@@ -268,6 +268,339 @@ def eval(self, sess, feed_dict, natoms):
...
```
Comment (njzjz) [inline deepmd/loss/ener.py:271]: This class is not tested

File: deepmd/entrypoints/test.py [hunk 23]
```diff
@@ -306,6 +306,26 @@ def test_ener(
         av = ret[4]
         ae = ae.reshape([numb_test, -1])
         av = av.reshape([numb_test, -1])
+    if dp.get_ntypes_spin() != 0:
```
Comment (njzjz) [inline deepmd/entrypoints/test.py:309]: The changes to this file are not tested

File: source/api_cc/include/DeepPot.h [hunk 24]
```diff
@@ -472,6 +481,14 @@ class DeepPotModelDevi {
     assert(inited);
     return ntypes;
   };
+  /**
+   * @brief Get the number of types with spin.
+   * @return The number of types with spin.
+   **/
+  int numb_types_spin() const {
```
Comment (njzjz) [inline source/api_cc/include/DeepPot.h:488]: This new public method is not tested.

File: source/api_cc/include/DeepPot.h [hunk 25]
```diff
@@ -260,6 +260,14 @@ class DeepPot {
...
```
Comment (njzjz) [inline source/api_cc/include/DeepPot.h:267]: This new public method is not tested. They should have a test so they can be migrate to the C API

File: deepmd/descriptor/se_a.py [hunk 26]
```diff
@@ -160,6 +163,7 @@ def __init__(
         precision: str = "default",
         uniform_seed: bool = False,
         multi_task: bool = False,
+        spin: Spin = None,
```
Comment (njzjz) [inline deepmd/descriptor/se_a.py:166]:
```suggestion
        spin: Optional[Spin] = None,
```

File: deepmd/descriptor/se_a.py [hunk 27]
```diff
@@ -542,6 +555,20 @@ def build(
...
+                t_virtual_len = tf.constant(
+                    self.spin.virtual_len,
+                    name="virtual_len",
+                    dtype=GLOBAL_TF_FLOAT_PRECISION,
+                )
+                t_spin_norm = tf.constant(
+                    self.spin.spin_norm,
+                    name="spin_norm",
+                    dtype=GLOBAL_TF_FLOAT_PRECISION,
+                )
```
Comment (njzjz) [inline deepmd/descriptor/se_a.py:571]: These codes are not directly related to the descriptor.
I suggest moving these lines to `Spin.build`, and call `Spin.build` in `EnerModel`.

File: deepmd/descriptor/se_a.py [hunk 28]
```diff
@@ -186,6 +190,15 @@ def __init__(
             self.exclude_types.add((tt[1], tt[0]))
         self.set_davg_zero = set_davg_zero
         self.type_one_side = type_one_side
+        self.spin = spin
+
+        # extend sel_a for spin system
+        if self.spin is not None:
+            self.ntypes_spin = self.spin.get_ntypes_spin()
+            self.sel_a_spin = self.sel_a[: self.ntypes_spin]
+            self.sel_a.extend(self.sel_a_spin)
```
Comment (njzjz) [inline deepmd/descriptor/se_a.py:199]: consider applying them to `se_r` and `hybrid`

File: deepmd/fit/ener.py [hunk 29]
```diff
@@ -477,6 +483,20 @@ def build(
             if self.aparam_inv_std is None:
                 self.aparam_inv_std = 1.0
 
+        ntypes_atom = self.ntypes - self.ntypes_spin
+        if self.spin is not None:
+            for type_i in range(self.ntypes):
+                if type_i >= ntypes_atom:
+                    break
```
Comment (njzjz) [inline deepmd/fit/ener.py:490]:
```suggestion
            for type_i in range(ntypes_atom):
```

**Review by njzjz — CHANGES_REQUESTED** [review]

File: source/api_cc/include/DeepPot.h (hunk 24 above)
Comment (hztttt) [reply source/api_cc/include/DeepPot.h:488]: Could you give me an example about how to test methods?

File: deepmd/loss/ener.py (hunk 22 above)
Comment (hztttt) [reply deepmd/loss/ener.py:271]: Could you give me a test example about `EnerStdLoss `？

File: deepmd/entrypoints/test.py (hunk 23 above)
Comment (hztttt) [reply deepmd/entrypoints/test.py:309]: Could you give me an example?

File: source/api_cc/include/DeepPot.h (hunk 24 above)
Comment (njzjz) [reply source/api_cc/include/DeepPot.h:488]: https://github.com/deepmodeling/deepmd-kit/tree/devel/source/api_cc/tests

File: deepmd/loss/ener.py (hunk 22 above)
Comment (njzjz) [reply deepmd/loss/ener.py:271]: You can use dp train to train a model then it will be covered

File: deepmd/entrypoints/test.py (hunk 23 above)
Comment (hztttt) [reply deepmd/entrypoints/test.py:322]: We have added ut.

File: deepmd/descriptor/se_a.py (hunk 28 above)
Comment (hztttt) [reply deepmd/descriptor/se_a.py:199]: We have applied  them to `se_r` and `hybrid`.

File: source/api_cc/include/DeepPot.h (hunk 24 above)
Comment (yangtengleo) [reply source/api_cc/include/DeepPot.h:488]: Could you offer us a workflow for a new test_*.cc in this directory? We don't have much experience in GTest framework.

**Review by njzjz — APPROVED** [review]

**Comment by njzjz** [comment]
Please resolve conflicts then we can merge.

File: examples/spin/se_e2_a/input.json
Comment (njzjz) [inline examples/spin/se_e2_a/input.json:1]: Please add this file to `source/tests/test_examples.py`.

File: source/api_cc/include/DeepPot.h (hunk 24 above)
Comment (hztttt) [reply source/api_cc/include/DeepPot.h:488]: OK.

**Review by njzjz — APPROVED** [review]

File: deepmd/model/ener.py [hunk 30]
```diff
@@ -187,8 +193,18 @@ def build(
...
+            type_spin = self.spin.build(
+                reuse=reuse,
+                suffix=suffix,
+            )
         input_dict["atype"] = atype_
 
+        # # spin if any
+        # if self.spin is not None:
+        #     input_dict['spin'] = self.spin
+
```
Comment (wanghan-iapcm) [inline deepmd/model/ener.py:207]: please remove the comment

File: deepmd/model/ener.py [hunk 31]
```diff
@@ -248,9 +264,15 @@ def build(
...
-        )
+        if self.spin is None:
+            energy_raw = tf.reshape(
+                energy_raw, [-1, natoms[0]], name="o_atom_energy" + suffix
+            )
+        else:
+            nloc_atom = tf.reduce_sum(natoms[2 : 2 + len(self.spin.use_spin)])
+            energy_raw = tf.reshape(
+                energy_raw, [-1, nloc_atom], name="o_atom_energy" + suffix
+            )
```
Comment (wanghan-iapcm) [inline deepmd/model/ener.py:275]:
```suggestion
        nloc_atom = natoms[0] if self.spin is None else tf.reduce_sum(natoms[2 : 2 + len(self.spin.use_spin)]) 
        energy_raw = tf.reshape(
            energy_raw, [-1, nloc_atom], name="o_atom_energy" + suffix
        )
```

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: deepmd/model/ener.py (hunk 30 above)
Comment (hztttt) [reply deepmd/model/ener.py]: OK.

**Review by wanghan-iapcm — APPROVED** [review]
